Mathlib/Computability/NFA.lean

```diff
@@ -104,6 +152,159 @@ def accepts : Language α := {x | ∃ S ∈ M.accept, S ∈ M.eval x}
 theorem mem_accepts {x : List α} : x ∈ M.accepts ↔ ∃ S ∈ M.accept, S ∈ M.evalFrom M.start x := by
   rfl
 
+variable (M) in
+/-- `M.Path` represents a concrete path through the NFA from a start state to an end state
+for a particular word.
+
+Note that due to the non-deterministic nature of the automata, the can be more than one `Path`
+for a given word.
+
+Also note that this is `Type` and not a `Prop`, so that we can speak about the properties
+of a particular `Path`, such as the set of states visited along the way (defined as `Path.supp`). -/
+inductive Path : σ → σ → List α → Type (max u v)
+  | nil (s : σ) : Path s s []
+  | cons (t s u : σ) (a : α) (x : List α) :
+      t ∈ M.step s a → Path t u x → Path s u (a :: x)
+
+/-- Knowing that a path must exist, construct one particular example. -/
+noncomputable def path_of_evalFrom {s t : σ} {x : List α} (h : t ∈ M.evalFrom {s} x) :
+    M.Path s t x :=
+  match x with
+  | [] =>
+    have h : s = t := by simp at h; tauto
+    h ▸ Path.nil s
+  | a :: x =>
+    have h : ∃ s' ∈ M.step s a, t ∈ M.evalFrom {s'} x :=
+      by rw [evalFrom_cons, mem_evalFrom_iff_exists, stepSet_singleton] at h; exact h
+    have ⟨s', h₁, h₂⟩ := Classical.indefiniteDescription _ h
+    Path.cons s' _ _ _ _ h₁ (path_of_evalFrom h₂)
+
+theorem evalFrom_of_path {s t : σ} {x : List α} (h : M.Path s t x) : t ∈ M.evalFrom {s} x := by
+  induction h with
+  | nil => simp
+  | cons s' s t a x h₁ h₂ ih =>
+    simp [M.mem_evalFrom_iff_exists (S := M.step _ _)]
+    tauto
+
+/-- Knowing that the NFA accepts a word, choose a concrete start state, path, and accept state. -/
+noncomputable def path_of_accepts {x : List α} (h : x ∈ M.accepts) :
+    (s t : σ) × M.Path s t x ×' s ∈ M.start ∧ t ∈ M.accept :=
+  have ⟨t, ht, h⟩ := Classical.indefiniteDescription _ (M.mem_accepts.1 h)
+  have ⟨s, hs, h⟩ := Classical.indefiniteDescription _ (M.mem_evalFrom_iff_exists.1 h)
+  ⟨s, t, M.path_of_evalFrom h, hs, ht⟩
+
+protected theorem Path.accepts {s t : σ} {x : List α} (p : M.Path s t x)
+    (hs : s ∈ M.start) (ht : t ∈ M.accept) : x ∈ M.accepts := by
+  apply M.evalFrom_of_path at p
+  simp [mem_accepts, M.mem_evalFrom_iff_exists (S := M.start)]
+  tauto
+
+/-- Append two paths. -/
+def Path.append {s t u : σ} {x y : List α} (p : M.Path s t x) (q : M.Path t u y) :
+    M.Path s u (x ++ y) :=
+  match p with
+  | nil _ => q
+  | cons s' _ _ _ _ h p' =>
+    cons s' _ _ _ _ h (p'.append q)
+
+/-- Split `p` at the point just after it has matched `x`. -/
+def Path.splitAfter {s u : σ} {y : List α} (x : List α) (p : M.Path s u (x ++ y)) :
+    (t : σ) × M.Path s t x × M.Path t u y :=
```

Comment:
... and here, etc...

Reply:
I don't believe this is the case here, `M.Path` is `Type`-valued, and together with `x` it's enough to infer everything.

Reply:
Still, it is very non-idiomatic having *mathematical* functions (just saying this because metaprogramming functions are quite exactly the opposite) return terms of a sigma or product type. They would instead be split into several functions. See eg [`SimpleGraph.Walk.takeUntil`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkDecomp.html#SimpleGraph.Walk.takeUntil) and [`SimpleGraph.Walk.dropUntil`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkDecomp.html#SimpleGraph.Walk.dropUntil)

Reply:
I suppose this could be split into two functions, but the product type would still be necessary in order to provide the `t` and be able to actually write down the return type.

Reply:
Why can't `t` be its own function? Eg here isn't it `x.getLastD s`?